File: GHCAA.Mobile/lib/features/scholarships/scholarship_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api/api_client.dart';
import 'scholarship_models.dart';

export 'scholarship_models.dart';

final scholarshipServiceProvider = Provider<ScholarshipService>(
  (ref) => ScholarshipService(ref.read(dioProvider)),
);

class ScholarshipService {
  final Dio _dio;

  ScholarshipService(this._dio);

  Future<List<ScholarshipFund>> getPublicFunds() async {
    final response = await _dio.get('/scholarships/public/funds');
    return (response.data as List)
        .map((item) => ScholarshipFund.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<List<ScholarshipCall>> getPublicCalls() async {
    final response = await _dio.get('/scholarships/public/calls');
    return (response.data as List)
        .map((item) => ScholarshipCall.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<ScholarshipApplication> submitApplication(
    int callId,
    CreateScholarshipApplication application,
  ) async {
    final response = await _dio.post(
      '/scholarships/calls/$callId/applications',
      data: application.toJson(),
    );
    return ScholarshipApplication.fromJson(
      response.data as Map<String, dynamic>,
    );
  }

  Future<ScholarshipStatus> getStatus(
    String referenceCode,
    String email,
  ) async {
    final response = await _dio.get(
      '/scholarships/status/${Uri.encodeComponent(referenceCode)}',
      queryParameters: {'email': email},
    );
    return ScholarshipStatus.fromJson(response.data as Map<String, dynamic>);
  }
}

File: GHCAA.Web/src/app/core/services/scholarship.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_ENDPOINTS } from '../constants/app.constants';
import {
    CreateScholarshipApplication, ScholarshipApplication, ScholarshipCall, ScholarshipFund, ScholarshipStatus
} from '../models/business.models';

@Injectable({ providedIn: 'root' })
export class ScholarshipService {
    private http = inject(HttpClient);
    private apiUrl = API_ENDPOINTS.SCHOLARSHIPS;

    getPublicFunds(): Observable<ScholarshipFund[]> {
        return this.http.get<ScholarshipFund[]>(`${this.apiUrl}/public/funds`);
    }

    getPublicCalls(): Observable<ScholarshipCall[]> {
        return this.http.get<ScholarshipCall[]>(`${this.apiUrl}/public/calls`);
    }

    submitApplication(callId: number, dto: CreateScholarshipApplication): Observable<ScholarshipApplication> {
        return this.http.post<ScholarshipApplication>(`${this.apiUrl}/calls/${callId}/applications`, dto);
    }

    getStatus(referenceCode: string, email: string): Observable<ScholarshipStatus> {
        const params = new HttpParams().set('email', email);
        return this.http.get<ScholarshipStatus>(`${this.apiUrl}/status/${encodeURIComponent(referenceCode)}`, { params });
    }
}
